src/server/pkg/backoff/retry.go

```diff
@@ -49,19 +109,18 @@ func RetryNotify(operation Operation, b BackOff, notify Notify) error {
 			}
 		}
 
-		time.Sleep(next)
-	}
-}
-
-// RetryUntilCancel is the same as RetryNotify, except that it will not retry if
-// the given context is canceled.
-func RetryUntilCancel(ctx context.Context, operation Operation, b BackOff, notify Notify) error {
-	return RetryNotify(operation, b, func(err error, d time.Duration) error {
 		select {
+		case <-time.After(next):
+			if ctx.Err() != nil {
+				// Corner case: if 'b' is ZeroBackOff, and 'ctx' is cancelled inside of
+				// notify(), then it's random which case is chosen, and the cancellation
+				// won't necessarily prevent opration() from running again. We can't
+				// guarantee that 'operation' is called with a live context, but this
```

Comment:
We _could_ guarantee that it's a live context if we do a non-blocking wait on `ctx.Done()` - I think that would be the right thing to do for this corner case.

Reply:
I mean, yeah, it would always be a race condition in a multi-threaded model, but I feel like having a race condition when there is a single thread of execution is silly.